### PR TITLE
Reactor integration

### PR DESCRIPTION
**Pull request by mykolaKhlopyk**
Integrate reactor to rest module. Nats controllers will be updated in future. Disable compile nats controllers, nats bpp and tests by using gradle exclude.

### CODE REVIEW

File: rest/src/main/kotlin/systems/ajax/englishstudytelegrambot/repository/LibraryRepository.kt [hunk 1]
```diff
@@ -1,89 +1,86 @@
...
+class LibraryRepositoryImpl(val mongoTemplate: ReactiveMongoTemplate) : LibraryRepository {
+
+    override fun saveNewLibrary(nameOfNewLibrary: String, telegramUserId: String): Mono<Library> =
+        mongoTemplate.insert(Library(name = nameOfNewLibrary, ownerId = telegramUserId))
+
+    override fun getAllLibraries(): Flux<Library> = mongoTemplate.findAll<Library>()
+
+    override fun deleteLibrary(libraryId: ObjectId): Mono<Library> =
+        Mono
+            .zip(
```
Comment (ajax-shulepa-a) [inline rest/src/main/kotlin/systems/ajax/englishstudytelegrambot/repository/LibraryRepository.kt]:
```suggestion
        Mono.zip(
```
require less tabs

File: rest/src/main/kotlin/systems/ajax/englishstudytelegrambot/service/LibraryService.kt [hunk 2]
```diff
@@ -1,35 +1,54 @@
...
-    }
-
-    override fun getAllWordsFromLibrary(libraryName: String, telegramUserId: String): List<Word> {
-        val libraryId = libraryRepository.getLibraryIdByLibraryNameAndTelegramUserId(libraryName, telegramUserId)
-        return libraryRepository.getAllWordsFromLibrary(libraryId)
-    }
+    override fun createNewLibrary(nameOfNewLibrary: String, telegramUserId: String): Mono<LibraryDtoResponse> =
+        libraryRepository
+            .saveNewLibrary(nameOfNewLibrary, telegramUserId)
+            .onErrorMap{LibraryWithTheSameNameForUserAlreadyExistException()}
```
Comment (ajax-shulepa-a) [inline rest/src/main/kotlin/systems/ajax/englishstudytelegrambot/service/LibraryService.kt]:
```suggestion
    override fun createNewLibrary(nameOfNewLibrary: String, telegramUserId: String): Mono<LibraryDtoResponse> =
        libraryRepository
            .saveNewLibrary(nameOfNewLibrary, telegramUserId)
            .onErrorMap{ LibraryWithTheSameNameForUserAlreadyExistException() }
```
If you don't pass any argument to exception, consider making it object, so you won't create the same instance each time
Alternatively, pass original error as source

File: rest/src/main/kotlin/systems/ajax/englishstudytelegrambot/service/WordService.kt [hunk 3]
```diff
@@ -25,37 +52,118 @@ class WordServiceImpl(
...
+                libraryName,
+                telegramUserId,
+                createWordDtoRequest.spelling
+            )
+            .switchIfEmpty(
+                checkWordOrLibraryIsMissing(libraryName, telegramUserId)
+            )
+            .flatMap { wordId ->
+                wordRepository.updateWordTranslating(wordId, createWordDtoRequest.translate)
+            }
```
Comment (ajax-shulepa-a) [inline rest/src/main/kotlin/systems/ajax/englishstudytelegrambot/service/WordService.kt]:
```suggestion
            .switchIfEmpty(checkWordOrLibraryIsMissing(libraryName, telegramUserId))
            .flatMap { wordRepository.updateWordTranslating(it, createWordDtoRequest.translate) }
```
[nit] I prefer having reactor operator on one line if possible, it's easier to read and understand what's happening

**Review by ajax-shulepa-a — APPROVED** [review]

File: rest/src/main/kotlin/systems/ajax/englishstudytelegrambot/external/source/ExternalWordSource.kt [hunk 4]
```diff
@@ -1,39 +1,25 @@
...
+    ): Mono<String> =
+        webClient.get()
+            .uri(
+                link, wordSpelling, wordnikProperties.tokenKey
+            )
+            .retrieve()
+            .bodyToFlux(T::class.java)
+            .next()
+            .map(GettingPartOfAdditionalInfoAboutWord::partOfAdditionalInfoAboutWord)
+            .onErrorReturn("Missing")
```
Comment (ajax-gnatiuk-s) [inline rest/src/main/kotlin/systems/ajax/englishstudytelegrambot/external/source/ExternalWordSource.kt:29]:
```kotlin
.doOnError {
    log.error("additional info for word {} wasn't found", wordSpelling)
}
```
?

File: rest/src/main/kotlin/systems/ajax/englishstudytelegrambot/external/source/ExternalWordSource.kt [hunk 5]
```diff
@@ -1,39 +1,25 @@
...
-        val log = LoggerFactory.getLogger(this::class.java)
-    }
+    ): Mono<String> =
+        webClient.get()
+            .uri(
+                link, wordSpelling, wordnikProperties.tokenKey
+            )
+            .retrieve()
+            .bodyToFlux(T::class.java)
+            .next()
```
Comment (ajax-gnatiuk-s) [inline rest/src/main/kotlin/systems/ajax/englishstudytelegrambot/external/source/ExternalWordSource.kt]: First you convert the body to flux, then you just take one item.

Maybe it's better to use `.toEntity<T>()` instead

Comment (mykolaKhlopyk) [reply rest/src/main/kotlin/systems/ajax/englishstudytelegrambot/external/source/ExternalWordSource.kt]: I depend on response from wordnik api. For every request (audio, pronunciation..) it returns array of answers, but i want save only first. So when i try bodyToMono or toEntity it throws 
com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize value of type `systems.ajax.englishstudytelegrambot.dto.external.api.PronunciationOfWordResponse` from Array value (token `JsonToken.START_ARRAY`)
i dont need to work with  whole Array so I came to this solution

File: rest/src/main/kotlin/systems/ajax/englishstudytelegrambot/external/source/ExternalWordSource.kt (hunk 4 above)
Comment (mykolaKhlopyk) [reply rest/src/main/kotlin/systems/ajax/englishstudytelegrambot/external/source/ExternalWordSource.kt:29]: sometime wordnik doesnt return answer and in that case i put Missing like an answer

Comment (mykolaKhlopyk) [reply rest/src/main/kotlin/systems/ajax/englishstudytelegrambot/external/source/ExternalWordSource.kt:29]: but if you mean log when error happens, i agree and add

**Pull request merged**